### ISSUE TITLE
Speaker attribution and voice classification

### ISSUE THREAD

**Issue by div5yesh** [mozilla/DeepSpeech#2169]
Enhancement: Speaker attribution and voice classification

- identify and label speakers against voices in the training set.
- cluster similar voices.

**Comment by kdavis-mozilla** [comment]
We can consider this, but generally we're narrowly focused on STT.

**Comment by div5yesh** [comment]
Transcription that also label the speakers in the audio for each word or dialogue duration can have useful applications and would relate to metadata for STT.

So maybe having a `contrib` repo, that is an extension to STT which will host features that are related to speech in general. But would result in a powerful DeepSpeech library to cater all kinds of speech related problems.

What do you think?

**Comment by kdavis-mozilla** [comment]
@div5yesh A `contrib` repo is a reasonable idea. But the we still have to consider the bandwidth required for review and tests of `contrib` code and how`contrib` is updated across non-backward compatible releases.

What's your take @reuben and @lissyx?

**Comment by lissyx** [comment]
I have a hard time figuring out exactly how those pieces would have to stick together. IMHO, my experience with such kind of contrib repo is really mixed feelings: often broken, badly maintained. Provides poor user / dev experience, generates frustrations. I understand the need for the feature, but it requires extending the API. How would a contrib repo, in the end, should be integrated to provide that ?

**Comment by lissyx** [comment]
> I believe, having an API to do deep analysis of the speech itself to extract information (not just text) could be more useful.

We're not saying it's not useful :)


> I think extending the API would be reasonable.

Well, we have a `Metadata` exposed struct that you might be able to extend and experiment with if you are interested.

**Comment by rhamnett** [comment]
I'd also be interested in this

**Comment by shashankpr** [comment]
It is an interesting enhancement and there are quite many existing work related to this as well. One of the notable one (which I am using it as a separate fork in my org) is this one: https://github.com/resemble-ai/Resemblyzer

I would be happy to contribute to the feature of speaker identification and classification after I make myself with the DeepSpeech's codebase

**Comment by lissyx** [comment]
FTR, I've come accross this work https://github.com/ina-foss/inaSpeechSegmenter that might be useful in this context.

**Comment by diego-fustes** [comment]
I think that Speaker Diarization would be the most useful here as it does not require any training data. The X-Vector model is one of the most powerful implementations, see the [Kaldi model](https://kaldi-asr.org/models/m6)

**Comment by Tortoise17** [comment]
How to use it if you can give an example. as input wav file?

**Comment by shravanshetty1** [comment]
This may or may not help - https://github.com/tyiannak/pyAudioAnalysis
